Hashtable.c

```diff
@@ -88,32 +90,94 @@ size_t Hashtable_count(const Hashtable* this) {
+   2592277, 3110741, 3732887, 4479463, 5375371, 6450413, 7740517, 9288589,
+   11146307, 13375573, 16050689, 19260817, 23112977, 27735583, 33282701,
+   39939233, 47927081, 57512503, 69014987, 82818011, 99381577, 119257891,
+   143109469, 171731387, 206077643, 247293161, 296751781, 356102141, 427322587,
+   512787097, 615344489, 738413383, 886096061, 1063315271, 1275978331,
+   1531174013, 1837408799, 2204890543UL, 2645868653UL, 3175042391UL,
+   3810050851UL,
+   /* on 32-bit make sure we do not return primes not fitting in size_t */
+#if SIZE_MAX > 4294967295ULL
+   4572061027ULL, 5486473229ULL, 6583767889ULL, 7900521449ULL, 9480625733ULL,
+   /* Largest possible size should be 13652101063ULL == GROWTH_RATE((UINT32_MAX/3)*4)
+      we include some larger values in case the above math is wrong */
+   11376750877ULL, 13652101063ULL, 16382521261ULL, 19659025513ULL, 23590830631ULL,
+#endif
 };
 
 static size_t nextPrime(size_t n) {
-   /* on 32-bit make sure we do not return primes not fitting in size_t */
-   for (size_t i = 0; i < ARRAYSIZE(OEISprimes) && OEISprimes[i] < SIZE_MAX; i++) {
-      if (n <= OEISprimes[i])
-         return OEISprimes[i];
+   for (size_t i = 0; i < ARRAYSIZE(primes); i++) {
+      if (n < primes[i]) {
+         return primes[i];
+      }
    }
 
    CRT_fatalError("Hashtable: no prime found");
 }
 
+/* USABLE_FRACTION is the maximum hash map load.
+ * Currently set to 2/3 capacity.
+ *
+ * Testing indicates that the median and average probe length
+ * increases significantly after 2/3 of the hash map capacity.
+ *
+ * load = {size,capacity} / items
+ *
+ * | load | probe max | probe avg | probe median |
+ * | ---- | --------- | --------- | ------------ |
+ * | 0.60 |     90.58 |     20.19 |         0.65 |
+ * | 0.65 |    167.00 |     37.07 |         1.58 |
+ * | 0.70 |    230.54 |     61.40 |        15.54 |
+ * | 0.75 |    287.00 |     85.23 |        26.15 |
+ * | 0.80 |    287.00 |     94.71 |        55.93 |
+ *
+ */
+#define USABLE_FRACTION(n) (((n) << 1)/3)
+
+/* SHRINK_THRESHOLD is number of items at with the hash map should shrink.
+ * Currently set to 1/4 of the USABLE_FRACTION, which is ~13% of the total
+ * hash map size.
+ */
+#define SHRINK_THRESHOLD(n) (USABLE_FRACTION((n)) / 4)
+
+/* GROWTH_RATE. Growth rate upon hitting maximum load.
+ * Currently set to items*3.
+ * This means that hashes double in size when growing without deletions,
+ * but have more head room when the number of deletions is on a par with the
+ * number of insertions.
+ */
+#define GROWTH_RATE(h) ((h)->items*3)
+
+static inline bool Hashtable_shouldResize(const Hashtable* this) {
+   /* grow table */
+   return this->items >= USABLE_FRACTION(this->size) ||
+      /* shrink table */
+      (this->size > MIN_TABLE_SIZE && this->items <= SHRINK_THRESHOLD(this->size));
+}
+
 Hashtable* Hashtable_new(size_t size, bool owner) {
+   assert(MIN_TABLE_SIZE == primes[0]);
    Hashtable* this;
 
    this = xMalloc(sizeof(Hashtable));
```

Comment:
If we're touching this code anyway, can also just join this into one line:

```suggestion
   Hashtable* this = xMalloc(sizeof(Hashtable));
```